crates/prune/prune/src/pruner.rs

```diff
@@ -169,81 +173,69 @@ impl<DB: Database> Pruner<DB> {
-                );
-
-                let segment_start = Instant::now();
-                let previous_checkpoint = provider.get_prune_checkpoint(segment.segment())?;
-                let output = segment.prune(
-                    provider,
-                    PruneInput { previous_checkpoint, to_block, limiter: limiter.clone() },
-                )?;
-                if let Some(checkpoint) = output.checkpoint {
-                    segment
-                        .save_checkpoint(provider, checkpoint.as_prune_checkpoint(prune_mode))?;
-                }
-                self.metrics
-                    .get_prune_segment_metrics(segment.segment())
-                    .duration_seconds
-                    .record(segment_start.elapsed());
-                if let Some(highest_pruned_block) =
-                    output.checkpoint.and_then(|checkpoint| checkpoint.block_number)
-                {
-                    self.metrics
-                        .get_prune_segment_metrics(segment.segment())
-                        .highest_pruned_block
-                        .set(highest_pruned_block as f64);
-                }
+        let initial_segments_len = self.segments.len();
+        let mut pruned_segments = Vec::with_capacity(self.segments.len());
+
+        // Prune other segments, taking them from the front of the queue
+        while let Some(segment) = self.segments.pop_front() {
+            let segment_name = segment.segment();
 
-                progress = output.progress;
-
-                debug!(
-                    target: "pruner",
-                    segment = ?segment.segment(),
-                    ?purpose,
-                    %to_block,
-                    ?prune_mode,
-                    %output.pruned,
-                    "Segment pruning finished"
-                );
-
-                if output.pruned > 0 {
-                    limiter.increment_deleted_entries_count_by(output.pruned);
-                    pruned += output.pruned;
-                    stats.insert(segment.segment(), (output.progress, output.pruned));
+            let result = self.prune_segment(
+                provider,
+                tip_block_number,
+                limiter,
+                &*segment,
+                PrunePurpose::User,
+            );
+            pruned_segments.push(segment);
+
+            match result? {
+                ControlFlow::Continue(output) => {
+                    progress = output.progress;
+
+                    if output.pruned > 0 {
+                        limiter.increment_deleted_entries_count_by(output.pruned);
+                        pruned += output.pruned;
+                        stats.insert(segment_name, (output.progress, output.pruned));
+                    }
                 }
-            } else {
-                debug!(target: "pruner", segment = ?segment.segment(), ?purpose, "Nothing to prune for the segment");
+                ControlFlow::LimitReached => break,
```

Comment:
this will loop until time limit reached, even when there are no segments to prune, a waste of resources

Reply:
when the limit is reached, this https://github.com/paradigmxyz/reth/blob/2cd1505b29c0d0ab0103b30ed55c73b384b7ac65/crates/prune/prune/src/pruner.rs#L280-L282 will be triggered and we will do `break` in the match clause you commented on

Reply:
ah wait, you described a different issue, why it will loop if there are no segments?